Add (int/int32/nativeint)_of_string_opt primitives

bikallem opened the pull request:
`int_of_string_opt`, `Int32.of_string_opt` and `Nativeint.of_string_opt` are currently implemented using the `try int_of_string with ...` code pattern. 

I was investigating the implementation of these functions and saw a TODO note `(* TODO: provide this directly as a non-raising primitive. *)` - in https://github.com/ocaml/ocaml/blob/trunk/stdlib/stdlib.ml#L271.

The existing code is functionally correct. However, I believe the current non-raising version in this PR is slightly more performant  due to the absence of exceptions. 

@alainfrisch pinging you since I believe you created this note. :)

EDIT: Does this PR need a CHANGES file entry?

bikallem committed:
Add (int/int32/nativeint)_of_string_opt primitive

nojb commented:
Since this PR is motivated by performance, can we have some benchmarks?

jberdine reviewed (commented):
The code change looks good to me. I read it for correctness and behavior preservation, and left one style comment inline. I have not checked the bootstrap.

I do not know if changing `val` to `external` in an exposed stdlib interface qualifies as a breaking change.

Benchmarks would be welcome, but this change is an improvement irrespectively since it makes these `_opt` functions safe to use in exception handlers without disturbing the current backtrace.

nojb commented:
> it makes these `_opt` functions safe to use in exception handlers without disturbing the current backtrace.

Good point.

jberdine commented on runtime/ints.c:
```diff
@@ -136,7 +141,20 @@ CAMLprim value caml_int_compare(value v1, value v2)
 CAMLprim value caml_int_of_string(value s)
 {
-    return Val_long(parse_intnat(s, 8 * sizeof(value) - 1, INT_ERRMSG));
+  intnat res;
+  if (0 == parse_intnat(s, 8 * sizeof(value) - 1, &res)) {
+    return Val_long(res);
+  }
+  caml_failwith(INT_ERRMSG);
```
Style comment: I would test for `0 !=` and reorder the cases in these functions. It is more common and, I think, easier to read to have early returns or exits for the error cases, and let the successful code path reach the end of the function.

bikallem replied:
This is a good idea. Done.

bikallem committed:
update error handling

bikallem commented:
> Since this PR is motivated by performance, can we have some benchmarks?

A micro benchmark of the code using `exn` to generate `None` vs `None` being generated via the primitive.

```
╭─────────────────────────────────┬───────────────────────────┬───────────────────────────┬───────────────────────────╮
│name                             │  major-allocated          │  minor-allocated          │  monotonic-clock          │
├─────────────────────────────────┼───────────────────────────┼───────────────────────────┼───────────────────────────┤
│  int_of_string_opt (exn/None)   │             0.0000 mjw/run│             0.0000 mnw/run│             49.5076 ns/run│
│  int_of_string_opt (prim/None)  │             0.0000 mjw/run│             0.0000 mnw/run│             11.6431 ns/run│
│  int_of_string_opt (exn/Some)   │             0.0000 mjw/run│             0.0000 mnw/run│             15.5363 ns/run│
│  int_of_string_opt (prim/Some)  │             0.0000 mjw/run│             0.3425 mnw/run│             15.7976 ns/run│
╰─────────────────────────────────┴───────────────────────────┴───────────────────────────┴───────────────────────────╯
```
If you want, I can post the code used to benchmark in a gist. 

EDIT: Here is the gist: https://gist.github.com/bikallem/7f61a1bf955691369e3e196aca34df91

jberdine commented:
Thanks for the style update. I rechecked the code: still good. I am in favor or merging this change.

(For the record it would be good to report the compilation command used for the benchmark, and in particular whether flambda was enabled or not, since that can influence removal of unnecessary allocations significantly.)

gasche reviewed (approved):
I agree that this sounds like a reasonable improvement, and the code is fairly clean.

gasche commented:
I have a question on the benchmark results: why does the exn/Some line not show any allocation at all?

bikallem commented:
> I have a question on the benchmark results: why does the exn/Some line not show any allocation at all?

Baffled me too. IIUC the allocation of the both should be about the same, right?

bikallem commented:
> Thanks for the style update. I rechecked the code: still good. I am in favor or merging this change.
>
Thanls. 
 
> (For the record it would be good to report the compilation command used for the benchmark, and in particular whether flambda was enabled or not, since that can influence removal of unnecessary allocations significantly.)

Sure. Tried both flambda and non flambda, the result was identical. The `dune` file that I have used for the above benchmark is below:
```
(executable 
  (name opt_bench)
  (modules opt_bench)
  (libraries bechamel notty.unix bechamel-notty))
```
```dune build opt_bench.ml --profile=release``` 
``` _build/default/opt_bench.exe```

lthls commented:
> why does the exn/Some line not show any allocation at all?

My guess is that the allocation in OCaml is removed by the optimiser, while the allocation in C cannot be removed.
In general you should be careful of benchmarks that drop the results of their computations: you might be benchmarking the compiler's ability to track the flow of values rather than the cost of the operation itself.
You can try storing the result of `int_of_string_opt` in an array in a tight loop, that will likely give you some useful measurements too.

gasche commented:
In any case, I would prefer if we could get performance numbers without suspicious unexplained questions. @bikallem would you be able to tweak the benchmark using @lthls' suggestions to see if the results make more sense?

dbuenzli commented:
> You can try storing the result of `int_of_string_opt` in an array in a tight loop,

For my own understanding, isn't it enough to `Sys.opaque_identity` the value ?

lthls commented:
> For my own understanding, isn't it enough to `Sys.opaque_identity` the value ?

As far as the OCaml compiler is concerned, both should work as well, but with the array version you introduce a computational dependency that is even harder to remove for the processor, so that makes the test a little more robust in my opinion. The downside is that you have a relatively expensive array store in the loop, so it becomes much harder to spot the difference between the two versions.

Although I expect that the initial benchmark actually shows something very relevant: in many cases, the user will call `int_of_string_opt` and immediately match on the option. The benchmark accurately shows that in this case, the `exn` version can see its allocation removed while the `prim` version cannot.

I'm also surprised by the huge difference between the two versions in the `None` case: raising an exception in OCaml is supposed to be fast, but here it looks like it's taking several times the amount of actual work being done. I'd like to see, if possible, the impact of enabling backtraces (call `Printexc.record_backtrace` before both tests with the appropriate boolean).

bikallem commented:
> In any case, I would prefer if we could get performance numbers without suspicious unexplained questions. @bikallem would you be able to tweak the benchmark using @lthls' suggestions to see if the results make more sense?

I have updated the benchmark to use both suggestions(@dbuenzli  and @lthls ). Here are the new benchmark numbers:

non-flambda:
```
╭────────────────────────────────────────┬───────────────────────────┬───────────────────────────┬───────────────────────────╮
│name                                    │  major-allocated          │  minor-allocated          │  monotonic-clock          │
├────────────────────────────────────────┼───────────────────────────┼───────────────────────────┼───────────────────────────┤
│  int_of_string_opt (arr/exn/None)      │           694.1373 mjw/run│          4015.1382 mnw/run│          40605.4892 ns/run│
│  int_of_string_opt (arr/exn/Some)      │          2309.6144 mjw/run│          1536.5714 mnw/run│          31979.3432 ns/run│
│  int_of_string_opt (arr/prim/None)     │           729.1693 mjw/run│             0.0000 mnw/run│          13136.6340 ns/run│
│  int_of_string_opt (arr/prim/Some)     │          2323.0129 mjw/run│          1545.4727 mnw/run│          36706.7150 ns/run│
│  int_of_string_opt (opaque/exn/None)   │             0.0000 mjw/run│             0.0000 mnw/run│             50.9688 ns/run│
│  int_of_string_opt (opaque/exn/Some)   │             0.0000 mjw/run│             0.0000 mnw/run│             14.7243 ns/run│
│  int_of_string_opt (opaque/prim/None)  │             0.0000 mjw/run│             0.0000 mnw/run│             16.2874 ns/run│
│  int_of_string_opt (opaque/prim/Some)  │             0.0000 mjw/run│             0.0000 mnw/run│             16.3910 ns/run│
╰────────────────────────────────────────┴───────────────────────────┴───────────────────────────┴───────────────────────────╯

```

flambda:
```
╭────────────────────────────────────────┬───────────────────────────┬───────────────────────────┬───────────────────────────╮
│name                                    │  major-allocated          │  minor-allocated          │  monotonic-clock          │
├────────────────────────────────────────┼───────────────────────────┼───────────────────────────┼───────────────────────────┤
│  int_of_string_opt (arr/exn/None)      │           691.8039 mjw/run│          4126.4061 mnw/run│          40440.3674 ns/run│
│  int_of_string_opt (arr/exn/Some)      │          2346.2340 mjw/run│          1560.8980 mnw/run│          31688.7912 ns/run│
│  int_of_string_opt (arr/prim/None)     │           733.8620 mjw/run│             0.0000 mnw/run│          13067.5352 ns/run│
│  int_of_string_opt (arr/prim/Some)     │          2356.1900 mjw/run│          1567.5108 mnw/run│          35988.0864 ns/run│
│  int_of_string_opt (opaque/exn/None)   │             0.0000 mjw/run│             0.0000 mnw/run│             49.3112 ns/run│
│  int_of_string_opt (opaque/exn/Some)   │             0.0000 mjw/run│             0.0000 mnw/run│             13.5260 ns/run│
│  int_of_string_opt (opaque/prim/None)  │             0.0000 mjw/run│             0.0000 mnw/run│             14.9611 ns/run│
│  int_of_string_opt (opaque/prim/Some)  │             0.0000 mjw/run│             0.2822 mnw/run│             16.0294 ns/run│
╰────────────────────────────────────────┴───────────────────────────┴───────────────────────────┴───────────────────────────╯
```

I have updated the gist too.

lthls commented:
I'm not sure I can trust benchmark numbers reporting zero allocations in the `Some` case, even with `opaque_identity`...

xavierleroy commented:
This is the same proposal as #10381, and it's going to be rejected for the same reason: it is faster to allocate the Some constructor in OCaml than in C, and that's the most common case.  So, the current implementation is fine.

Can we PLEASE remove this TODO comment in stdlib/stdlib.ml ?  There is nothing TODO here, and it has already misled two contributors into submitting PRs.

alainfrisch commented:
Apologies for having written this TODO in the first place.

The argument about not destroying the current exception trace suggests there might be something to do in order to avoid using exception, while still allocating the `Some` in OCaml.  In #10381, it was suggested to use a low-level hack to let the C primitive notifies the OCaml side about the error case, e.g. returning a block, and checking for that special return value on the OCaml side (but this would force using something like `Obj.tag` which might end up being even slower).

lthls commented:
> it was suggested to use a low-level hack to let the C primitive notifies the OCaml side about the error case, e.g. returning a block, and checking for that special return value on the OCaml side

@gasche's proposal for unboxed constructors might allow us to implement this suggestion in a type-safe way in the future (one unboxed constructor for the integer returned, one normal constructor for the error case).

nojb commented:
See https://github.com/ocaml/ocaml/pull/12952 for the comment. I think the PR can be closed. @bikallem: sorry about that!

bikallem commented:
No problem.
